### DISCUSSION TITLE
Using Casadi with C++ running on hardware in real-time

### DISCUSSION THREAD

**Discussion by Kamal-a-1** [zehuilu/Tutorial-on-CasADi-with-CPP#3] [Q&A]
Hi,
Thank you very much for this tutorial, that is really helpful.

I am wondering if it is possible to use Casadi functions in a cpp code and then compile it for any hardware (microcontroller for example) in order to run it in real-time. Is this possible or would there be compilation errors preventing it from being loaded into a hardware?

And in general, any other suggestions on how to effectively use Casadi solvers ('ipopt' for example) on a real-time hardware?

Thanks very much :)

**Comment by zehuilu** [comment]
Hi,

For the first question, I would suggest checking the entire Casadi community to see if there are any examples of embedded systems.

For the second question, to call IPOPT in real-time efficiently, typically we need to generate analytical gradient and Hessian and pass them to IPOPT. This can be done by CasADi automatically if we do not manually change the setting. Secondly, compiling the solver offline and only evaluating it at run-time would be efficient. If you need to solve an optimization with some changing parameters, like model predictive control where the initial state keeps changing, you need to set up a parameter argument "p" in CasADi solvers, just like "x". In this case, you don't have to re-construct a new optimization object at run-time. I will add this example later.

In general, IPOPT is one powerful computational tool for solving optimization problems. Depending on the problem structure, we might be able to use other methods like SQP (sequential quadratic programming). Determining which solver to use is problem-dependent and requires domain knowledge of the particular application. This is still an active research field.

Good luck :)
